DeeperFunc/Deeper.swift

```diff
@@ -0,0 +1,75 @@
+//
+//  Deeper.swift
+//  DeeperFunc
+//
+//  Created by Ilya Puchka on 26/10/2017.
+//  Copyright © 2017 Ilya Puchka. All rights reserved.
+//
+
+import Foundation
+
+public enum Either<A, B> {
+    case left(A), right(B)
+}
+
+public typealias RouteComponents = (path: [String], query: [String: String])
+
+public protocol PatternState {}
+public protocol ClosedPatternState: PatternState {} // pattern is complete
+public protocol OpenPatternState: PatternState {} // pattern requires subsequent pattern to become closed
+
+public protocol ClosedPathPatternState: ClosedPatternState {}
+public enum Path: ClosedPathPatternState {} // somewhere in the path
+public enum Query: ClosedPatternState {} // query started
+
+public struct RoutePattern<A/*pattern type*/, S: PatternState> {
+    public let parse: Parser<A> // parses path components exctracting underlying type of pattern
+    public let print: Printer<A> // converts pattern with passed in value to template component
```

Comment:
`print` is actually not that useful in context of deep links as we don't need to produce urls in code, so probably can be dropped. Though it can be used for application wide navigation, so that `openURL` is used always instead of passing manually created intents around.